src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,297 @@
+
+const API_TIMEOUT = 15000
+
+export const uploadToGitDB = async (
+  log: Logger,
+  request: RequestBuilder,
+  git: simpleGit.SimpleGit,
+  dryRun: boolean
+) => {
+  let repoURL
+  try {
+    repoURL = await getRepoURL(git)
+    log.debug(`Syncing repository ${repoURL}`)
+  } catch (err) {
+    log.warn(`Failed getting repository URL: ${err}`)
+    throw err
+  }
+
+  await unshallowRepositoryWhenNeeded(log, git)
+
+  let latestCommits
+  try {
+    latestCommits = await getLatestLocalCommits(git)
+    if (latestCommits.length === 0) {
+      log.debug('No local commits found.')
+
+      return
+    }
+    log.debug(`${latestCommits.length} commits found, asking GitDB which ones are missing.`)
+  } catch (err) {
+    log.warn(`Failed getting local commits: ${err}`)
+    throw err
+  }
+
+  let commitsToExclude
+  try {
+    commitsToExclude = await getKnownCommits(log, request, repoURL, latestCommits)
+    log.debug(`${commitsToExclude.length} commits already in GitDB.`)
+  } catch (err) {
+    log.warn(`Failed getting commits to exclude: ${err}`)
+    throw err
+  }
+
+  // Get the list of all objects (commits, trees) to upload. This list can be quite long
+  // so quite memory intensive (multiple MBs).
+  let objectsToUpload
+  try {
+    objectsToUpload = await getObjectsToUpload(git, commitsToExclude)
+    log.debug(`${objectsToUpload.length} objects to upload.`)
+  } catch (err) {
+    log.warn(`Failed getting objects to upload: ${err}`)
+    throw err
+  }
+
+  let packfiles
+  try {
+    packfiles = generatePackFilesForCommits(log, objectsToUpload)
+    log.debug(`${packfiles.length} packfiles generated.`)
+  } catch (err) {
+    log.warn(`Failed generating packfiles: ${err}`)
+    throw err
+  }
+
+  if (dryRun) {
+    log.debug(`Dry-run enabled, not uploading anything.`)
+
+    return
+  }
+  log.debug(`Uploading packfiles...`)
+  try {
+    await uploadPackfiles(log, request, repoURL, latestCommits[0], packfiles)
+    log.debug('Successfully uploaded packfiles.')
+  } catch (err) {
+    log.warn(`Failed to upload packfiles: ${err}`)
+    throw err
+  }
+}
+
+const getRepoURL = gitRemote
+
+const getLatestLocalCommits = async (git: simpleGit.SimpleGit) => {
+  const logResult = await git.log(['-n 1000', '--since="1 month ago"'])
+
+  return logResult.all.map((c) => c.hash)
+}
+
+const unshallowRepositoryWhenNeeded = async (log: Logger, git: simpleGit.SimpleGit) => {
+  const isShallow = (await git.revparse('--is-shallow-repository')) === 'true'
+  if (isShallow) {
+    log.info('[unshallow] Git repository is a shallow clone., unshallowing it...')
+    log.info('[unshallow] Setting remote.origin.partialclonefilter to "blob:none" to avoid fetching file content')
+    await git.addConfig('remote.origin.partialclonefilter', 'blob:none')
+    log.info('[unshallow] Running git fetch --shallow-since="1 month ago" --update-shallow --refetch')
+    await git.fetch(['--shallow-since="1 month ago"', '--update-shallow', '--refetch'])
```

Comment:
Do you know if `1 month ago` is always 30d, or if we're in March we could end up with only 28d of data ? Would it be safer to put `30 days ago` here ?

Reply:
I'm pretty sure the answer is we'll have only 28d of data. I don't think that's too bad as we expect this limit to be useful only when uploading the repo for the first time, and having 28d or 30d doesn't really change much. I will keep the current behavior as this is also what is implemented in the tracer. cc @edrevo FYI